sdk/core/Azure.Core/src/Shared/ClientDiagnostics.cs

```diff
@@ -2,52 +2,154 @@
 // Licensed under the MIT License.
 
 using System;
+using System.Collections.Generic;
 using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
 using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
 
 #nullable enable
 
 namespace Azure.Core.Pipeline
 {
-#pragma warning disable CA1001 // Types that own disposable fields should be disposable
-    internal sealed class ClientDiagnostics
-#pragma warning restore CA1001 // Types that own disposable fields should be disposable
+    internal sealed class ClientDiagnostics: DiagnosticScopeFactory
     {
-        private readonly string? _resourceProviderNamespace;
-        private readonly DiagnosticListener? _source;
+        private const string DefaultMessage = "Service request failed.";
 
-        public bool IsActivityEnabled { get;  }
+        private readonly HttpMessageSanitizer _sanitizer;
+        public ClientDiagnostics(ClientOptions options) : base(
+            options.GetType().Namespace,
```

Comment:
Given this is pseudo-public with a bunch of service client assemblies, should we do some amount of null checks to avoid ugly NullReferenceExceptions?

Reply:
Is that needed with `#nullable enable` being invoked and the parameter not being decorated as allowing `null`?

Reply:
Clients should be validating the options objects way before calling into `ClientDiagnostics`, I'm not a fan of excessive internal defensive null checks.

>Is that needed with #nullable enable being invoked and the parameter not being decorated as allowing null?

If the consuming project doesn't have nullable attributes enabled the warning won't be emitted.

Reply:
@jsquire nullability only affects compile-time checks currently, and if projects don't use them (like the service projects) nothing checks it.